File: verifiers/typescript/src/tiers/tier1Pbt.ts
// tier1Pbt.ts — fast-check / @fast-check/vitest driver.
//
// Pins (May 2026): fast-check 4.7.0, @fast-check/vitest 0.4.1.
//
// Tier 1 spec (docs/01-architecture/verifier-pipeline.md):
// - "Verifier requires authored property tests covering the changed
//   function's invariants. Runs them at >=10,000 iterations."
// - Wall-clock budget: 5 min default, 15 min max.
//
// How we run:
// 1. Discover *.test.ts / *.spec.ts files in the diff that import
//    fast-check.
// 2. Spawn `vitest run <files>` with VITEST_FAST_CHECK_NUM_RUNS=10000
//    in the environment. Tests written with @fast-check/vitest can
//    also pin numRuns inline via `it.prop([gen], { numRuns: 10000 })`
//    — we set the env so existing tests at the default numRuns are
//    automatically scaled up to the Crucible minimum.
// 3. Parse vitest's JSON reporter output to detect counterexamples
//    (fast-check prints "Counterexample:" in the failure message).
//
// If the diff has no property tests at all, we emit `failed` with a
// finding — Tier 1 is mandatory when the dispatcher routes us here, so
// "no PBTs" is a real failure, not a tool_unavailable.

import { spawn } from "node:child_process";
import { mkdtemp, readFile, rm, writeFile } from "node:fs/promises";
import path from "node:path";
import os from "node:os";

import { classifyDiff } from "../diff.js";
import {
  finalizeTestReport,
  newTestReport,
  type Counterexample,
  type Finding,
  type PBTStats,
  type TestReport,
  type VerificationRequest,
} from "../schema.js";

export const IT_PROP_MIN_RUNS = 10_000;

export interface Tier1Options {
  cwd?: string;
  vitestBin?: string;
  numRuns?: number;
  wallClockSeconds?: number;
  /**
   * Pre-supplied vitest JSON report content. Used by tests to bypass
   * the subprocess spawn. When set, `testFiles` is ignored.
   */
  preSuppliedReport?: string;
  /** Override the file-discovery step (test-only). */
  testFilesOverride?: readonly string[];
}

// ---- vitest JSON reporter shape (v2.x subset) ----------------------------

interface VitestAssertionResult {
  status: "passed" | "failed" | "skipped" | "pending" | "todo";
  fullName: string;
  failureMessages?: string[];
}

interface VitestTestResult {
  assertionResults: VitestAssertionResult[];
  status: "passed" | "failed";
}

interface VitestReport {
  numTotalTestSuites: number;
  numPassedTestSuites: number;
  numFailedTestSuites: number;
  numTotalTests: number;
  numPassedTests: number;
  numFailedTests: number;
  testResults: VitestTestResult[];
}

// ---- fast-check counterexample parsing -----------------------------------

const COUNTEREXAMPLE_RE =
  /Counterexample:\s*(?<shrunk>[\s\S]*?)(?:\nShrunk\s+(?<shrinks>\d+)\s+time\(s\))?\nGot(?:\s+error)?:\s*(?<got>[\s\S]*?)(?:\nseed:\s*(?<seed>-?\d+))?(?:\n|$)/i;

export function parseCounterexample(
  property: string,
  failureMessage: string,
): Counterexample | null {
  const m = COUNTEREXAMPLE_RE.exec(failureMessage);
  if (!m || !m.groups) {
    return null;
  }
  const shrunk = (m.groups["shrunk"] ?? "").trim();
  const got = (m.groups["got"] ?? "").trim();
  const seed = m.groups["seed"];
  const ce: Counterexample = {
    property,
    shrunk,
    stack_hint: got.slice(0, 500),
  };
  if (seed !== undefined && seed !== "") {
    ce.seed = seed;
  }
  return ce;
}

// ---- File discovery ------------------------------------------------------

async function findPbtFiles(
  cwd: string,
  candidates: readonly string[],
): Promise<string[]> {
  const out: string[] = [];
  for (const rel of candidates) {
    const abs = path.isAbsolute(rel) ? rel : path.join(cwd, rel);
    try {
      const src = await readFile(abs, "utf8");
      if (/from\s+["']fast-check["']/.test(src) ||
          /from\s+["']@fast-check\/vitest["']/.test(src) ||
          /require\(["']fast-check["']\)/.test(src)) {
        out.push(rel);
      }
    } catch {
      // missing on disk — skip
    }
  }
  return out;
}

// ---- Subprocess driver ---------------------------------------------------

function runVitest(
  bin: string,
  cwd: string,
  files: readonly string[],
  reportPath: string,
  numRuns: number,
  wallClockSeconds: number,
): Promise<{ exitCode: number; stderr: string }> {
  return new Promise((resolve, reject) => {
    const args = [
      "run",
      "--reporter=json",
      `--outputFile=${reportPath}`,
      ...files,
    ];
    const child = spawn(bin, args, {
      cwd,
      stdio: ["ignore", "pipe", "pipe"],
      env: {
        ...process.env,
        // @fast-check/vitest honours this env var for its `it.prop` runs.
        VITEST_FAST_CHECK_NUM_RUNS: String(numRuns),
        // fast-check's vanilla `fc.assert` also honours these defaults
        // when no inline numRuns is provided.
        FAST_CHECK_NUM_RUNS: String(numRuns),
        FORCE_COLOR: "0",
      },
    });
    let stderr = "";
    let killed = false;
    const timer = setTimeout(
      () => {
        killed = true;
        child.kill("SIGKILL");
      },
      Math.max(1, wallClockSeconds) * 1000,
    );
    child.stderr?.on("data", (b: Buffer) => {
      stderr += b.toString("utf8");
      process.stderr.write(b);
    });
    child.stdout?.on("data", (b: Buffer) => {
      // vitest JSON reporter writes to outputFile; anything on stdout
      // is incidental — forward to stderr to keep our protocol clean.
      process.stderr.write(b);
    });
    child.on("error", (err) => {
      clearTimeout(timer);
      reject(err);
    });
    child.on("exit", (code) => {
      clearTimeout(timer);
      if (killed) {
        resolve({
          exitCode: 124,
          stderr: stderr + "\n[crucible: killed by wall-clock budget]",
        });
        return;
      }
      resolve({ exitCode: code ?? 1, stderr });
    });
  });
}

// ---- Public entrypoint ---------------------------------------------------

export async function runTier1Pbt(
  req: VerificationRequest,
  opts: Tier1Options = {},
): Promise<TestReport> {
  const startedAt = new Date();
  const cwd = opts.cwd ?? process.cwd();
  const numRuns = opts.numRuns ?? IT_PROP_MIN_RUNS;
  const wallClockSeconds =
    opts.wallClockSeconds ??
    Math.max(300, Math.min(900, req.budget.wall_clock_cap_seconds || 300));

  const baseReport = newTestReport({
    task_id: req.task_id,
    diff_hash: req.diff.base_sha ?? req.base_sha,
    tier: "tier_1_pbt",
    framework: "fast-check-4.7.0+@fast-check/vitest-0.4.1",
    reporter_id: "crucible-verify-typescript",
    reporter_version: "2026.06.0-phase4",
    wall_clock_budget_seconds: wallClockSeconds,
    started_at: startedAt,
  });

  // Find candidate test files.
  const cls = classifyDiff(req.diff);
  const candidates =
    opts.testFilesOverride !== undefined
      ? Array.from(opts.testFilesOverride)
      : cls.tests.map((f) => f.path);

  const pbtFiles =
    opts.preSuppliedReport !== undefined
      ? candidates
      : await findPbtFiles(cwd, candidates);

  if (pbtFiles.length === 0 && opts.preSuppliedReport === undefined) {
    const finding: Finding = {
      category: "pbt_missing",
      severity: "error",
      detail:
        "No property-based test files found in diff. Tier 1 requires at least one *.test.ts that imports fast-check.",
    };
    return finalizeTestReport(
      {
        ...baseReport,
        verdict: "failed",
        passed: false,
        pbt: {
          iterations: 0,
          iterations_min: IT_PROP_MIN_RUNS,
        },
        findings: [finding],
      },
      new Date(),
    );
  }

  // Run vitest.
  let raw: string;
  if (opts.preSuppliedReport !== undefined) {
    raw = opts.preSuppliedReport;
  } else {
    const tmp = await mkdtemp(path.join(os.tmpdir(), "crucible-tier1-"));
    const reportPath = path.join(tmp, "vitest.json");
    try {
      const result = await runVitest(
        opts.vitestBin ?? "vitest",
        cwd,
        pbtFiles,
        reportPath,
        numRuns,
        wallClockSeconds,
      );
      try {
        raw = await readFile(reportPath, "utf8");
      } catch {
        return finalizeTestReport(
          {
            ...baseReport,
            verdict: "failed",
            passed: false,
            error: `vitest report missing (exit ${result.exitCode}). stderr tail: ${result.stderr.slice(-1500)}`,
          },
          new Date(),
        );
      }
    } catch (err) {
      return finalizeTestReport(
        {
          ...baseReport,
          verdict: "tool_unavailable",
          passed: false,
          error: `vitest spawn failed: ${(err as Error).message}`,
        },
        new Date(),
      );
    } finally {
      // Best-effort cleanup.
      void rm(tmp, { recursive: true, force: true }).catch(() => {});
    }
  }

  let report: VitestReport;
  try {
    report = JSON.parse(raw) as VitestReport;
  } catch (err) {
    return finalizeTestReport(
      {
        ...baseReport,
        verdict: "failed",
        passed: false,
        error: `vitest JSON report unparsable: ${(err as Error).message}`,
      },
      new Date(),
    );
  }

  // Collect counterexamples + property names.
  const counterexamples: Counterexample[] = [];
  const propertyNames: string[] = [];
  for (const tr of report.testResults ?? []) {
    for (const a of tr.assertionResults ?? []) {
      propertyNames.push(a.fullName);
      if (a.status === "failed") {
        const msg = (a.failureMessages ?? []).join("\n");
        const ce = parseCounterexample(a.fullName, msg);
        if (ce !== null) {
          counterexamples.push(ce);
        } else {
          counterexamples.push({
            property: a.fullName,
            shrunk: msg.slice(0, 500),
            stack_hint: msg.slice(0, 1000),
          });
        }
      }
    }
  }

  // Each test invocation runs `numRuns` iterations under @fast-check/vitest.
  const iterations = (report.numTotalTests || 0) * numRuns;
  const passed =
    report.numFailedTests === 0 &&
    iterations >= IT_PROP_MIN_RUNS &&
    counterexamples.length === 0;

  const pbt: PBTStats = {
    iterations,
    iterations_min: IT_PROP_MIN_RUNS,
    properties: propertyNames,
    ...(counterexamples.length > 0 ? { counterexamples } : {}),
  };

  const findings: Finding[] = counterexamples.slice(0, 50).map((c) => ({
    category: "property_failed",
    severity: "error" as const,
    detail: `Property "${c.property}" found counterexample: ${c.shrunk}`,
  }));

  return finalizeTestReport(
    {
      ...baseReport,
      pbt,
      findings,
      verdict: passed ? "passed" : "failed",
      passed,
    },
    new Date(),
  );
}

export const __testing__ = { parseCounterexample, findPbtFiles };

File: verifiers/rust/src/tiers/tier1_pbt.rs
//! Tier-1 property-based tests + fuzz.
//!
//! Pipeline:
//!
//! 1. Run `cargo test --tests` with `PROPTEST_CASES=10000` so proptest
//!    auto-runs at the Crucible-mandated iteration count.
//! 2. Scan stdout/stderr for `Falsifying example` lines (proptest's
//!    shrunk-counterexample marker) and `test ... panicked` lines.
//! 3. If a `fuzz/` directory exists in the working tree, locate the
//!    first fuzz target via `cargo fuzz list` and run
//!    `cargo fuzz run <target> -- -max_total_time=15`. Any crashes are
//!    recorded as `fuzz_crashes`.

use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::Instant;

use regex::Regex;

use crate::schema::{Counterexample, Finding, PbtStats, TestReport, Verdict, VerificationRequest};
use crate::tiers::locate;

/// Crucible mandate: proptest must run at ≥10,000 iterations.
pub const ITERATIONS_MIN: u64 = 10_000;

/// Maximum wall-clock for `cargo fuzz run` (seconds). Kept short to stay
/// within Tier-1's 5-minute default budget; the dispatcher's own
/// wall-clock cap is the upper bound.
pub const FUZZ_MAX_SECONDS: u64 = 15;

/// Entrypoint invoked by `tiers::dispatch`.
pub fn run(_req: &VerificationRequest, mut report: TestReport) -> TestReport {
    report.framework = "proptest+cargo-fuzz".to_string();
    let started = Instant::now();

    let Some(cargo) = locate("cargo") else {
        report.stamp_finished();
        return report.tool_unavailable("cargo not on PATH");
    };

    let mut stats = PbtStats {
        iterations: 0,
        iterations_min: ITERATIONS_MIN,
        ..PbtStats::default()
    };

    // ---- proptest pass -------------------------------------------------
    let test_output = Command::new(&cargo)
        .arg("test")
        .arg("--tests")
        .arg("--no-fail-fast")
        .env("PROPTEST_CASES", ITERATIONS_MIN.to_string())
        // Force colour off for predictable parsing.
        .env("CARGO_TERM_COLOR", "never")
        .output();

    let test_result = match test_output {
        Ok(o) => o,
        Err(e) => {
            report.stamp_finished();
            return report.tool_unavailable(format!("cargo test spawn: {e}"));
        }
    };

    let stdout = String::from_utf8_lossy(&test_result.stdout);
    let stderr = String::from_utf8_lossy(&test_result.stderr);

    let (props, counters) = parse_proptest_output(&stdout, &stderr);
    stats.properties = props;
    stats.iterations = if stats.properties.is_empty() {
        0
    } else {
        ITERATIONS_MIN * stats.properties.len() as u64
    };
    stats.counterexamples = counters;

    let pbt_passed = test_result.status.success() && stats.counterexamples.is_empty();

    // ---- fuzz pass -----------------------------------------------------
    // The fuzz directory lives at the workdir's root, not in `req`.
    let fuzz_dir = std::env::current_dir()
        .ok()
        .map(|d| d.join("fuzz"))
        .filter(|p| p.exists());

    if let Some(fdir) = fuzz_dir.as_ref() {
        match run_fuzz(&cargo, fdir) {
            Ok(fuzz) => {
                stats.fuzz_corpus_size = fuzz.corpus_size;
                stats.fuzz_new_seeds = fuzz.new_seeds;
                stats.fuzz_crashes = fuzz.crashes;
            }
            Err(e) => {
                eprintln!("crucible-verify-rust: cargo fuzz skipped: {e}");
            }
        }
    }

    report.findings = stats
        .counterexamples
        .iter()
        .map(|c| Finding {
            category: "property_failed".to_string(),
            severity: "error".to_string(),
            file: String::new(),
            line: 0,
            detail: format!("{}: {}", c.property, c.shrunk),
            suggested_fix: String::new(),
        })
        .collect();

    let passed = pbt_passed && stats.fuzz_crashes == 0;
    report.verdict = if passed {
        Verdict::Passed
    } else {
        Verdict::Failed
    };
    report.passed = passed;
    report.pbt = Some(stats);
    report.duration_seconds = started.elapsed().as_secs_f64();
    report.stamp_finished();
    report
}

#[derive(Debug, Default)]
struct FuzzOutcome {
    corpus_size: u64,
    new_seeds: u64,
    crashes: u64,
}

fn run_fuzz(cargo: &Path, fuzz_dir: &Path) -> anyhow::Result<FuzzOutcome> {
    // List fuzz targets first.
    let list = Command::new(cargo)
        .arg("fuzz")
        .arg("list")
        .current_dir(fuzz_dir.parent().unwrap_or(fuzz_dir))
        .output()?;
    let listed = String::from_utf8_lossy(&list.stdout);
    let target = listed
        .lines()
        .map(str::trim)
        .find(|l| !l.is_empty())
        .ok_or_else(|| anyhow::anyhow!("cargo fuzz list returned no targets"))?
        .to_string();

    let run = Command::new(cargo)
        .arg("fuzz")
        .arg("run")
        .arg(&target)
        .arg("--")
        .arg(format!("-max_total_time={FUZZ_MAX_SECONDS}"))
        .current_dir(fuzz_dir.parent().unwrap_or(fuzz_dir))
        .output()?;
    let stderr = String::from_utf8_lossy(&run.stderr);
    let crashes = u64::from(stderr.contains("crash-") || stderr.contains("ERROR:"));

    let corpus_dir: PathBuf = fuzz_dir.join("corpus").join(&target);
    let mut corpus_size = 0u64;
    if let Ok(rd) = std::fs::read_dir(&corpus_dir) {
        corpus_size = rd.filter_map(Result::ok).count() as u64;
    }
    Ok(FuzzOutcome {
        corpus_size,
        new_seeds: 0,
        crashes,
    })
}

/// Parse proptest output for the two markers that matter:
///   - `proptest: Modifying configuration: …` (kept for completeness)
///   - `Falsifying example: …` (shrunk counter-example)
///   - `thread 'test_name' panicked` (catch-all)
///
/// Returns (property names seen, counterexamples).
fn parse_proptest_output(stdout: &str, stderr: &str) -> (Vec<String>, Vec<Counterexample>) {
    // Pattern: "test foo::bar ..." that proptest annotates as a property.
    let test_name_re = Regex::new(r"(?m)^test\s+([A-Za-z0-9_:]+)\s+\.\.\.").expect("static regex");
    let falsifying_re = Regex::new(r"(?ms)Falsifying example for `([^`]+)`:\s*(.+?)(?:\n\n|\z)")
        .expect("static regex");
    let panic_re = Regex::new(r"thread '([^']+)' panicked at ([^\n]+)").expect("static regex");

    let mut properties = Vec::new();
    for cap in test_name_re.captures_iter(stdout) {
        let name = cap[1].to_string();
        if name.contains("prop_") || name.contains("property_") {
            properties.push(name);
        }
    }
    properties.sort();
    properties.dedup();

    let mut counters = Vec::new();
    for cap in falsifying_re.captures_iter(stdout) {
        counters.push(Counterexample {
            property: cap[1].to_string(),
            shrunk: cap[2].trim().to_string(),
            seed: String::new(),
            stack_hint: String::new(),
        });
    }
    for cap in falsifying_re.captures_iter(stderr) {
        counters.push(Counterexample {
            property: cap[1].to_string(),
            shrunk: cap[2].trim().to_string(),
            seed: String::new(),
            stack_hint: String::new(),
        });
    }
    for cap in panic_re.captures_iter(stderr) {
        // A panic without a matching `Falsifying example` line. Still
        // record it so the dispatcher sees the failure.
        if !counters.iter().any(|c| c.property == cap[1]) {
            counters.push(Counterexample {
                property: cap[1].to_string(),
                shrunk: cap[2].trim().to_string(),
                seed: String::new(),
                stack_hint: "panic".to_string(),
            });
        }
    }
    (properties, counters)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn captures_falsifying_example() {
        let stdout = "
test prop_addition_is_commutative ... FAILED
Falsifying example for `prop_addition_is_commutative`:
    a = 1
    b = 2

test result: FAILED. 0 passed; 1 failed
";
        let (props, counters) = parse_proptest_output(stdout, "");
        assert!(props.iter().any(|p| p.contains("prop_addition")));
        assert_eq!(counters.len(), 1);
        assert_eq!(counters[0].property, "prop_addition_is_commutative");
        assert!(counters[0].shrunk.contains("a = 1"));
    }

    #[test]
    fn captures_panic_without_falsifying_line() {
        let stderr = "thread 'tests::prop_sample' panicked at src/lib.rs:10:5\n";
        let (_, counters) = parse_proptest_output("", stderr);
        assert_eq!(counters.len(), 1);
        assert_eq!(counters[0].stack_hint, "panic");
    }
}
